File: streampark-flink/streampark-flink-kubernetes/src/test/scala/org/apache/streampark/flink/kubernetes/PodTemplateParserTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.kubernetes

import org.apache.streampark.common.util.Implicits._

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

// scalastyle:off println
class PodTemplateParserTest {

  @Test
  def testCompleteInitPodTemplate(): Unit = {
    val podTemplateExpect = Map(
      "" ->
        """apiVersion: v1
          |kind: Pod
          |metadata:
          |  name: pod-template
          |""".stripMargin,
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: jm-pod-template
        |""".stripMargin
        ->
          """apiVersion: v1
            |kind: Pod
            |metadata:
            |  name: jm-pod-template
            |""".stripMargin,
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: pod-template
        |spec:
        |""".stripMargin
        ->
          """apiVersion: v1
            |kind: Pod
            |metadata:
            |  name: pod-template
            |""".stripMargin,
      """apiVersion: v1
        |spec:
        |""".stripMargin
        ->
          """apiVersion: v1
            |kind: Pod
            |metadata:
            |  name: pod-template
            |""".stripMargin,
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: pod-template
        |spec:
        |  containers:
        |    - name: flink-main-container
        |      volumeMounts:
        |        - name: checkpoint-pvc
        |          mountPath: /opt/flink/checkpoints
        |        - name: savepoint-pvc
        |          mountPath: /opt/flink/savepoints
        |  volumes:
        |    - name: checkpoint-pvc
        |      persistentVolumeClaim:
        |        claimName: flink-checkpoint
        |    - name: savepoint-pvc
        |      persistentVolumeClaim:
        |        claimName: flink-savepoint
        |""".stripMargin
        ->
          """apiVersion: v1
            |kind: Pod
            |metadata:
            |  name: pod-template
            |spec:
            |  containers:
            |  - name: flink-main-container
            |    volumeMounts:
            |    - name: checkpoint-pvc
            |      mountPath: /opt/flink/checkpoints
            |    - name: savepoint-pvc
            |      mountPath: /opt/flink/savepoints
            |  volumes:
            |  - name: checkpoint-pvc
            |    persistentVolumeClaim:
            |      claimName: flink-checkpoint
            |  - name: savepoint-pvc
            |    persistentVolumeClaim:
            |      claimName: flink-savepoint
            |""".stripMargin)
    for (expect <- podTemplateExpect) {
      val res = PodTemplateParser.completeInitPodTemplate(expect._1)
      assertEquals(expect._2.trim, res.trim)
    }
  }

  @Test
  def testExtractHostAliasMapFromPodTemplate(): Unit = {
    val expected = Map(
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: pod-template
        |spec:
        |  containers:
        |  - name: flink-main-container
        |    volumeMounts:
        |    - name: checkpoint-pvc
        |      mountPath: /opt/flink/checkpoints
        |    - name: savepoint-pvc
        |      mountPath: /opt/flink/savepoints
        |  volumes:
        |  - name: checkpoint-pvc
        |    persistentVolumeClaim:
        |      claimName: flink-checkpoint
        |  - name: savepoint-pvc
        |    persistentVolumeClaim:
        |      claimName: flink-savepoint
        |  hostAliases:
        |  - ip: 192.168.3.114
        |    hostnames:
        |    - hdp01.assad.site
        |    - hdp01
        |  - ip: 192.168.3.116
        |    hostnames:
        |    - hdp03.assad.site
        |  - ip: 192.168.3.115
        |    hostnames:
        |    - hdp02.assad.site
        |    - hdp02
        |""".stripMargin
        ->
          Map(
            "hdp01.assad.site" -> "192.168.3.114",
            "hdp01" -> "192.168.3.114",
            "hdp03.assad.site" -> "192.168.3.116",
            "hdp02.assad.site" -> "192.168.3.115",
            "hdp02" -> "192.168.3.115"),
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: pod-template
        |spec:
        |  containers:
        |  - name: flink-main-container
        |    volumeMounts:
        |    - name: checkpoint-pvc
        |      mountPath: /opt/flink/checkpoints
        |    - name: savepoint-pvc
        |      mountPath: /opt/flink/savepoints
        |  volumes:
        |  - name: checkpoint-pvc
        |    persistentVolumeClaim:
        |      claimName: flink-checkpoint
        |  - name: savepoint-pvc
        |    persistentVolumeClaim:
        |      claimName: flink-savepoint
        |  hostAliases:
        |""".stripMargin
        ->
          Map(),
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: pod-template
        |spec:
        |  hostAliases:
        |  - ip:
        |    hostnames:
        |    - hdp01.assad.site
        |    - hdp01
        |  - ip: 192.168.3.116
        |    hostname:
        |    - hdp03.assad.site
        |  - ip: 192.168.3.115
        |    hostnames:
        |    - hdp02.assad.site
        |  - ip: 192.168.3.115
        |    hostname: hdp02.assad.site
        |""".stripMargin
        -> Map(),
      """apiVersion: v1
        |kind: Pod
        |metadata:
        |  name: pod-template
        |spec: 2333
        |""".stripMargin
        -> Map())

    for (expect <- expected) {
      val hostsMap = PodTemplateParser.extractHostAliasMap(expect._1)
      val map = expect._2
      assertEquals(hostsMap.size(), map.size)
      hostsMap.foreach(c => {
        assertEquals(c._2, map.get(c._1))
      })
    }
  }

}

File: streampark-flink/streampark-flink-kubernetes/src/test/java/org/apache/streampark/flink/kubernetes/PodTemplateParserTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.kubernetes;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

class PodTemplateParserTest {

    @Test
    void testCompleteInitPodTemplate() {
        Map<String, String> podTemplateExpect = new HashMap<>();
        podTemplateExpect.put(
            "",
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n");
        podTemplateExpect.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: jm-pod-template\n",
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: jm-pod-template\n");
        podTemplateExpect.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec:\n",
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n");
        podTemplateExpect.put(
            "apiVersion: v1\n" + "spec:\n",
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n");
        podTemplateExpect.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec:\n"
                + "  containers:\n"
                + "    - name: flink-main-container\n"
                + "      volumeMounts:\n"
                + "        - name: checkpoint-pvc\n"
                + "          mountPath: /opt/flink/checkpoints\n"
                + "        - name: savepoint-pvc\n"
                + "          mountPath: /opt/flink/savepoints\n"
                + "  volumes:\n"
                + "    - name: checkpoint-pvc\n"
                + "      persistentVolumeClaim:\n"
                + "        claimName: flink-checkpoint\n"
                + "    - name: savepoint-pvc\n"
                + "      persistentVolumeClaim:\n"
                + "        claimName: flink-savepoint\n",
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec:\n"
                + "  containers:\n"
                + "  - name: flink-main-container\n"
                + "    volumeMounts:\n"
                + "    - name: checkpoint-pvc\n"
                + "      mountPath: /opt/flink/checkpoints\n"
                + "    - name: savepoint-pvc\n"
                + "      mountPath: /opt/flink/savepoints\n"
                + "  volumes:\n"
                + "  - name: checkpoint-pvc\n"
                + "    persistentVolumeClaim:\n"
                + "      claimName: flink-checkpoint\n"
                + "  - name: savepoint-pvc\n"
                + "    persistentVolumeClaim:\n"
                + "      claimName: flink-savepoint\n");

        for (Map.Entry<String, String> expect : podTemplateExpect.entrySet()) {
            String res = PodTemplateParser.completeInitPodTemplate(expect.getKey());
            Assertions.assertEquals(expect.getValue().trim(), res.trim());
        }
    }

    @Test
    void testExtractHostAliasMapFromPodTemplate() {
        Map<String, Map<String, String>> expected = new HashMap<>();
        expected.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec:\n"
                + "  containers:\n"
                + "  - name: flink-main-container\n"
                + "    volumeMounts:\n"
                + "    - name: checkpoint-pvc\n"
                + "      mountPath: /opt/flink/checkpoints\n"
                + "    - name: savepoint-pvc\n"
                + "      mountPath: /opt/flink/savepoints\n"
                + "  volumes:\n"
                + "  - name: checkpoint-pvc\n"
                + "    persistentVolumeClaim:\n"
                + "      claimName: flink-checkpoint\n"
                + "  - name: savepoint-pvc\n"
                + "    persistentVolumeClaim:\n"
                + "      claimName: flink-savepoint\n"
                + "  hostAliases:\n"
                + "  - ip: 192.168.3.114\n"
                + "    hostnames:\n"
                + "    - hdp01.assad.site\n"
                + "    - hdp01\n"
                + "  - ip: 192.168.3.116\n"
                + "    hostnames:\n"
                + "    - hdp03.assad.site\n"
                + "  - ip: 192.168.3.115\n"
                + "    hostnames:\n"
                + "    - hdp02.assad.site\n"
                + "    - hdp02\n",
            new HashMap<String, String>() {

                {
                    put("hdp01.assad.site", "192.168.3.114");
                    put("hdp01", "192.168.3.114");
                    put("hdp03.assad.site", "192.168.3.116");
                    put("hdp02.assad.site", "192.168.3.115");
                    put("hdp02", "192.168.3.115");
                }
            });
        expected.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec:\n"
                + "  containers:\n"
                + "  - name: flink-main-container\n"
                + "    volumeMounts:\n"
                + "    - name: checkpoint-pvc\n"
                + "      mountPath: /opt/flink/checkpoints\n"
                + "    - name: savepoint-pvc\n"
                + "      mountPath: /opt/flink/savepoints\n"
                + "  volumes:\n"
                + "  - name: checkpoint-pvc\n"
                + "    persistentVolumeClaim:\n"
                + "      claimName: flink-checkpoint\n"
                + "  - name: savepoint-pvc\n"
                + "    persistentVolumeClaim:\n"
                + "      claimName: flink-savepoint\n"
                + "  hostAliases:\n",
            new HashMap<>());
        expected.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec:\n"
                + "  hostAliases:\n"
                + "  - ip:\n"
                + "    hostnames:\n"
                + "    - hdp01.assad.site\n"
                + "    - hdp01\n"
                + "  - ip: 192.168.3.116\n"
                + "    hostname:\n"
                + "    - hdp03.assad.site\n"
                + "  - ip: 192.168.3.115\n"
                + "    hostnames:\n"
                + "    - hdp02.assad.site\n"
                + "  - ip: 192.168.3.115\n"
                + "    hostname: hdp02.assad.site\n",
            new HashMap<String, String>() {

                {
                    put("hdp02.assad.site", "192.168.3.115");
                }
            });
        expected.put(
            "apiVersion: v1\n"
                + "kind: Pod\n"
                + "metadata:\n"
                + "  name: pod-template\n"
                + "spec: 2333\n",
            new HashMap<>());

        for (Map.Entry<String, Map<String, String>> expect : expected.entrySet()) {
            Map<String, String> hostsMap = PodTemplateParser.extractHostAliasMap(expect.getKey());
            Map<String, String> map = expect.getValue();
            Assertions.assertEquals(map.size(), hostsMap.size());
            for (Map.Entry<String, String> entry : hostsMap.entrySet()) {
                Assertions.assertEquals(map.get(entry.getKey()), entry.getValue());
            }
        }
    }
}
